### PR TITLE
RustLexer improvements

### PR DESCRIPTION
**Pull request by jajakobyly**
Hi!

This PR fixes the buggy handling of string literals with invalid escapes, also some reformatting has been done in order to enhance readability.

Before merging I would like to dispel some doubts (and eventually fix them here):

- [x]  Why `'static` is treated as a separate token type? It makes no sense for me, special handling for this should be done in annotators imho.
- [x]  Why the set of possible suffixes for integers and floats is limited right in the lexer? The Rust Reference states, that any valid identifier can be a valid suffix. Again, I believe that validating them should be done using annotators as it's purely semantic thing, not syntactic. (String literal suffixes are already supported syntactically, though currently Rust does not allow any semantically)

### CODE REVIEW

File: toml [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit 5c8c55193c8c3eacfdbc819b9c13bc05782302db
+Subproject commit 20f655782626a02052971d87ff3ecfd6f01a8787
```
Comment (jajakobyly) [inline toml]: Ouch, this had unintentionally slipped, sorry. Unfortunately I don't know how to reverse it, any idea?

**Comment by matklad** [comment]
>Why 'static is treated as a separate token type? It makes no sense for me, special handling for this should be done in annotators imho.

Our parser and lexer are not exactly polished yet, there are a lot of things which can be done better. As for `STATIC_LIFITIME`, it exists because we've ported some parts of our parser from [parser-lalr.y](https://github.com/rust-lang/rust/blob/master/src/grammar/parser-lalr.y), which has `STATIC_LIFETIME`.  I believe it should be possible to get rid of this extra token. There should not be any interactions with the `static` keyword, right?

>Why the set of possible suffixes for integers and floats is limited right in the lexer?

It's just less code to restrict suffixes in lexer rather then creating a special annotator for this. But I agree that annotator is a better solution here, at least because of better error messages .

**Comment by matklad** [comment]
>Why the set of possible suffixes for integers and floats is limited right in the lexer?

Hm, in the reference suffixes for numbers are exhaustively listed: https://doc.rust-lang.org/reference.html#suffixes.

Arbitrary suffixes can interact badly with underscores. `1__________` is a valid literal without a suffix. So in the end it may be indeed simpler to leave suffix handling to lexer?

**Comment by jajakobyly** [comment]
> `1__________` is a valid literal without a suffix.

This case shouldn't be a problem. We only want to recognize that this is an integer literal, we don't want to know in `RustLexer` what is an integer itself and what is a suffix. Anyway, the rule:
```JFlex
[0-9] [0-9_]* {IDENTIFIER}?
```
should match this token without triggering the `{IDENTIFIER}?` part.

**Comment by matklad** [comment]
Ok, so removing `STATIC_LIFETIME`, `INT_SUFFIX`, `FLT_SUFFIX` and adding an annotator for suffixes seems like the right solution here.

**Comment by jajakobyly** [comment]
I have no idea why 23fe6cc breaks these tests (`RustPartialParsingTestCase > testFn` and `org.rust.lang.core.parser.RustPartialParsingTestCase > testIfExpr`)

**Comment by matklad** [comment]
They are test cases for parser error recovery, which include automatically generated error messages (based on FIRST an FOLLOW sets I think), so they often break when the parser is updated. 

You can just delete `fn.txt`, and run the test. First time it will fail, but will generate the new `fn.txt` which should be commuted (if has the same general shape, the previous `fn.txt`)

File: src/main/kotlin/org/rust/lang/core/grammar/rust.bnf [hunk 2]
```diff
@@ -1046,7 +1046,7 @@ private bool_lit ::= FALSE | TRUE
 // Lifetimes
 ///////////////////////////////////////////////////////////////////////////////////////////////////
 
-private lifetime ::= LIFETIME | STATIC_LIFETIME
+private lifetime ::= LIFETIME
```
Comment (matklad) [inline src/main/kotlin/org/rust/lang/core/grammar/rust.bnf]: This rule now seems unnecessary.

File: src/main/kotlin/org/rust/lang/core/lexer/RustLexer.flex [hunk 3]
```diff
@@ -133,142 +127,143 @@ UNICODE_ESCAPE = \\u\{{HEX_DIGIT}{1,6}\}
...
+  "/*"                            { yybegin(IN_BLOCK_COMMENT); yypushback(2); }
+  "//"                            { yybegin(IN_EOL_COMMENT);   yypushback(2); }
+
+  {IDENTIFIER}                    { return IDENTIFIER; }
 
   /* LITERALS */
 
-  {INT_LITERAL}                   { return RustTokenElementTypes.INTEGER_LITERAL; }
+  // Match 1f32 and 1f64 as floats, not integers (kinda hack)
+  {DEC_LITERAL} {FLT_SUFFIX}      { return FLOAT_LITERAL; }
```
Comment (jajakobyly) [inline src/main/kotlin/org/rust/lang/core/lexer/RustLexer.flex]: I know it's smelly, but I couldn't think of a better way of achieving this.

Comment (matklad) [inline src/main/kotlin/org/rust/lang/core/lexer/RustLexer.flex]: It's fine I think. Can't find a better solution either.

**Comment by matklad** [comment]
Do you want to add an annotator for integer suffixes in this PR as well?

**Comment by matklad** [comment]
The history will be a little nicer if you squash "upss" and "(T): "fix" broken tests" commits ;)

**Comment by jajakobyly** [comment]
I can make the annotator in a separate PR. Squash will be done tomorrow (AWK).

**Comment by jajakobyly** [comment]
Done

**Pull request merged**